src/methods/call.js

```diff
@@ -326,60 +327,114 @@ export function OrdinaryCallEvaluateBody(
       // 4. Return Completion{[[Type]]: return, [[Value]]: G, [[Target]]: empty}.
       return new ReturnCompletion(G, realm.currentLocation);
     } else {
-      // 1. Perform ? FunctionDeclarationInstantiation(F, argumentsList).
-      Functions.FunctionDeclarationInstantiation(realm, F, argumentsList);
-
-      // 2. Return the result of EvaluateBody of the parsed code that is the value of F's
-      //    [[ECMAScriptCode]] internal slot passing F as the argument.
-      let priorSavedCompletion = realm.savedCompletion;
+      if (!realm.useAbstractInterpretation || realm.pathConditions.length === 0) return normalCall();
+      let savedIsSelfRecursive = F.isSelfRecursive;
       try {
-        realm.savedCompletion = undefined;
+        F.isSelfRecursive = false;
+        let effects = realm.evaluateForEffects(guardedCall);
+        if (F.isSelfRecursive) {
+          AbstractValue.reportIntrospectionError(F, "call to function that calls itself");
+          throw new FatalError();
+          //todo: need to emit a specialized function that temporally captures the heap state at this point
+        } else {
+          realm.applyEffects(effects);
+          let c = effects[0];
+          return processResult(() => {
+            invariant(c instanceof Value || c instanceof AbruptCompletion);
+            return c;
+          });
+        }
+      } finally {
+        F.isSelfRecursive = savedIsSelfRecursive;
+      }
+
+      function guardedCall() {
+        let currentLocation = realm.currentLocation;
+        if (F.activeArguments !== undefined && F.activeArguments.has(currentLocation)) {
+          let [previousPathLength, previousArguments] = F.activeArguments.get(currentLocation);
+          if (realm.pathConditions.length > previousPathLength) {
+            invariant(previousArguments !== undefined);
+            // F is being called recursively while a call to it is still active
+            F.isSelfRecursive = true;
+            let widenedArgumentsList: Array<Value> = (Widen.widenValues(realm, previousArguments, argumentsList): any);
+            if (Widen.containsArraysOfValue(realm, previousArguments, widenedArgumentsList)) {
+              // Reached a fixed point. Executing this call will not add any knowledge
+              // about the effects of the original call.
+              return AbstractValue.createFromType(realm, Value, "widened return result");
```

Comment:
This is the place where an abstract recursive call is replaced with a Top value.